### PR TITLE
P2P car rental data model

### PR DESCRIPTION
**Pull request by ilarimikkonen**
This PR introduces a renting related data model, p2p car rental. It's developed in co-operation with Shareit Bloxcar Oy. The aim is to provide basic datamodel, which can be modified as needed.

### CODE REVIEW

File: specs/Renting/P2PCarRental/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# P2P Car Rental
+
```
Comment (jmcanterafonseca) [inline specs/Renting/P2PCarRental/README.md]: is this about renting or just another transportation option?

File: specs/Renting/P2PCarRental/doc/spec.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+document don't open up the usage.
+
+A JSON Schema corresponding to this data model can be found
+[here](../schema.json). All the following entries are mandatory.
+
+-   `id` : Unique identifier.
+
+-   `type` : Entity type. It must be equal to `P2PCarRental`.
+
+-   `vehicle_model` : The model of the vehicle omitting make, for example "159"
```
Comment (jmcanterafonseca) [inline specs/Renting/P2PCarRental/doc/spec.md]: please follow the general data model guidelines ... and the documentation template

File: specs/Renting/P2PCarRental/doc/spec.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# P2PCarRental
+
+## Description
+
+This entity contains a description of a peer to peer rental car.
```
Comment (jmcanterafonseca) [inline specs/Renting/P2PCarRental/doc/spec.md]: this is just a vehicle and we already have a Vehicle model ...

File: specs/Renting/P2PCarRental/README.md (hunk 1 above)
Comment (ilarimikkonen) [reply specs/Renting/P2PCarRental/README.md]: this is about renting

File: specs/Renting/P2PCarRental/doc/spec.md (hunk 3 above)
Comment (ilarimikkonen) [reply specs/Renting/P2PCarRental/doc/spec.md]: I checked the existing vehicle model and it did not seem to have all the stuff we need here. I'll check again to see how we can use that existing vehicle model. ANy suggestions?

**Comment by jmcanterafonseca** [comment]
why don't you just use the VehicleModel data model (proposing an extension) to capture the static properties of your vehicle and just define an Entity type named P2PCarRental with the specific renting needs?

**Comment by ilarimikkonen** [comment]
> why don't you just use the VehicleModel data model (proposing an extension) to capture the static properties of your vehicle and just define an Entity type named P2PCarRental with the specific renting needs?

Ok, sounds like a plan. How do you see the use of "id"? now the rental datamodel has an "id" and then the VehicleModel specifies "id" as mandatory. Should we use same ID for both, like "Rental-car-8f7ewf88ejjnf993jfU83HDJ" ?

**Comment by ilarimikkonen** [comment]
+ how do I reference this schema: https://github.com/FIWARE/dataModels/blob/master/specs/Transportation/Vehicle/VehicleModel/schema.json from my json schema?

 I don't understand how this: https://github.com/FIWARE/dataModels/blob/master/specs/Transportation/Vehicle/Vehicle/schema.json and this: https://github.com/FIWARE/dataModels/blob/master/specs/Transportation/Vehicle/VehicleModel/schema.json

play together

**Comment by jmcanterafonseca** [comment]
be careful, I'm talking about adding extra properties to VehicleModel not to create another VehicleModel schema. 

They don't play together they are a different kind of Entity. From Vehicle you point to a Vehicle Model.

**Comment by ilarimikkonen** [comment]
> be careful, I'm talking about adding extra properties to VehicleModel not to create another VehicleModel schema.
> 
> They don't play together they are a different kind of Entity. From Vehicle you point to a Vehicle Model.

Ok, got it. I'll get started so that I just include the current VehicleModel properties and then propose new ones bit later once I get this one sorted.

I notice that you do point from Vehicle to VehicleModel, but I dont understand how (this is my 1st data model, please be patient), since the ref seems to be like this:

"refVehicleModel": {"$ref": "https://fiware.github.io/dataModels/common-schema.json#/definitions/EntityIdentifierType" }

Is this meaning that in this case it points to a particular existing instance on VehicleModel?

**Comment by jmcanterafonseca** [comment]
build broken

https://travis-ci.org/FIWARE/dataModels/jobs/515171826

File: specs/Renting/P2PCarRental/doc/spec.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+-   `type` : Entity type. It must be equal to `P2PCarRental`.
+
+-   `vehicle_average_rating`: Float, range from 1-5. Peers can give a rating for the renting experience.
+	-   Attribute type: Float
+	-   Optional
+
+-   `timezone` : For example: "Europe/Helsinki".
+	-   Normative reference: [IANA](https://www.iana.org/time-zones) spec.
+
+-   `vehicle_information` : Follows [VehicleModel](https://github.com/FIWARE/dataModels/tree/master/specs/Transportation/Vehicle/VehicleModel) more details how to use the model are there. How ever, some of the semantics when using this datamodel are different. Here is a explanation how those are used in this model:
```
Comment (jmcanterafonseca) [inline specs/Renting/P2PCarRental/doc/spec.md]: be careful, I would reference a Vehicle Model but not inline it

**Comment by jmcanterafonseca** [comment]
> "refVehicleModel": {"$ref": "https://fiware.github.io/dataModels/common-schema.json#/definitions/EntityIdentifierType" }
> 
> Is this meaning that in this case it points to a particular existing instance on VehicleModel?

Yes it is. Essentially you are going to have an Entity of type `FleetVehicle`, category `carRental`, that is going to have a `refVehicleModel` pointing to a `VehicleModel` instance.

So, the approach is, make a PR that extends the VehicleModel data model. Then, make another PR that creates the FleetVehicle data model. 

So, you are far from the solution, yet :)

**Comment by jmcanterafonseca** [comment]
build is broken and feedback has not been taken into account, please check all comments again

**Comment by ilarimikkonen** [comment]
> build broken
> 
> https://travis-ci.org/FIWARE/dataModels/jobs/515171826

fix pushes, will hopefully resolve

**Comment by ilarimikkonen** [comment]
> > "refVehicleModel": {"$ref": "https://fiware.github.io/dataModels/common-schema.json#/definitions/EntityIdentifierType" }
> > Is this meaning that in this case it points to a particular existing instance on VehicleModel?
> 
> Yes it is. Essentially you are going to have an Entity of type `FleetVehicle`, category `carRental`, that is going to have a `refVehicleModel` pointing to a `VehicleModel` instance.
> 
> So, the approach is, make a PR that extends the VehicleModel data model. Then, make another PR that creates the FleetVehicle data model.
> 
> So, you are far from the solution, yet :)

I think we have some fundamental misunderstanding here. Question: Are all NGSI datamodels designed to work with Context Broker and only context broker?

**Comment by ilarimikkonen** [comment]
What I'm trying to also judge here is that there is an existing API producing JSON very close to the one model I initially created. Now if I try to reuse existing NGSI datamodel pieces, (Vehicle, VehicleModel) I need to rethink the whole model, as the existing API relates to *renting* and the fact that the object for rent happens to be a vehicle is coincidence. That is the reason I created new "Renting" section.

Is the fundamental NGSI philosophy that datamodel contributors should reuse as much existing datamodels as possible? What I mean here is that from effort perspective if there is an existing API producing JSON it is very easy to drop it to a tool and the tool spits out a matching schema. If the datamodel needs to be constructed from pieces of existing datamodels, the effort increases.

**Comment by jmcanterafonseca** [comment]
> Is the fundamental NGSI philosophy that datamodel contributors should reuse as much existing datamodels as possible? 

Yes, it is, we want coherency, we are welcoming a lot new contributions but they shall be aligned

> What I mean here is that from effort perspective if there is an existing API producing JSON it is very easy to drop it to a tool and the tool spits out a matching schema. If the datamodel needs to be constructed from pieces of existing datamodels, the effort increases.

Yes, but also your visibility and the reuse and impact of your work ...

**Comment by ilarimikkonen** [comment]
I'm planning on closing this pull request. Requirements have changed and I don't feel that this makes sense, as it looks like we do not have an active data source for this. @jmcanterafonseca if you still see value in this, I can proceed with this, but otherwise I'd close it.

**Comment by jmcanterafonseca** [comment]
You can close. Thanks

**Pull request closed without merging**